src/monitor/circuit_graph.rs

```diff
@@ -0,0 +1,379 @@
+}
+
+pub(super) enum NodeKind {
+    /// Regular operator.
+    Operator,
+    /// Root circuit or subcircuit.
+    Circuit {
+        iterative: bool,
+        children: HashMap<NodeId, Node>,
+        region: Region,
+    },
+    /// The input half of a strict operator.
+    StrictInput { output: NodeId },
+}
+
+/// A node in a circuit graph represents an operator or a circuit.
+pub(super) struct Node {
+    id: GlobalNodeId,
+    pub name: String,
+    #[allow(dead_code)]
+    pub region_id: RegionId,
+    pub kind: NodeKind,
+}
+
+impl Node {
+    pub(super) fn new(id: GlobalNodeId, name: &str, region_id: RegionId, kind: NodeKind) -> Self {
+        Self {
+            id,
+            name: name.to_string(),
+            region_id,
+            kind,
+        }
+    }
+
+    /// Lookup node in the subtree with the root in `self` by path.
+    fn node_ref<'a>(&self, mut path: slice::Iter<'a, NodeId>) -> Option<&Node> {
+        match path.next() {
+            None => Some(self),
+            Some(node_id) => match &self.kind {
+                NodeKind::Circuit { children, .. } => children.get(node_id)?.node_ref(path),
+                _ => None,
+            },
+        }
+    }
+
+    /// Lookup node in the subtree with the root in `self` by path.
+    fn node_mut<'a>(&mut self, mut path: slice::Iter<'a, NodeId>) -> Option<&mut Node> {
+        match path.next() {
+            None => Some(self),
+            Some(node_id) => match &mut self.kind {
+                NodeKind::Circuit { children, .. } => children.get_mut(node_id)?.node_mut(path),
+                _ => None,
+            },
+        }
+    }
+
+    /// `true` if `self` is a circuit node.
+    pub(super) fn is_circuit(&self) -> bool {
+        matches!(self.kind, NodeKind::Circuit { .. })
+    }
+
+    /// `true` if `self` is an iterative circuit (including the root
+    /// circuit).
+    pub(super) fn is_iterative(&self) -> bool {
+        matches!(
+            self.kind,
+            NodeKind::Circuit {
+                iterative: true,
+                ..
+            }
+        )
+    }
+
+    /// Returns children of `self` if `self` is a circuit.
+    pub(super) fn children(&self) -> Option<&HashMap<NodeId, Node>> {
+        if let NodeKind::Circuit { children, .. } = &self.kind {
+            Some(children)
+        } else {
+            None
+        }
+    }
+
+    /// Returns a mutable reference to the root region of `self` if
+    /// `self` is a circuit node.
+    pub(super) fn region_mut(&mut self) -> Option<&mut Region> {
+        if let NodeKind::Circuit { region, .. } = &mut self.kind {
+            Some(region)
+        } else {
+            None
+        }
+    }
+
+    /// Returns a reference to the root region of `self` if
+    /// `self` is a circuit node.
+    pub(super) fn region(&self) -> Option<&Region> {
+        if let NodeKind::Circuit { region, .. } = &self.kind {
+            Some(region)
+        } else {
+            None
+        }
+    }
```

Comment:
Convention is to name these methods as `.as_*`, e.g. `.as_children()` or `.as_mut_region()`